File: 2023-12-21/leetcode_110.py
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        self.isHeightBalanced = True

        def dfs(node, height = 0):
            if not node:
                return height
            height += 1
            left, right = dfs(node.left, height), dfs(node.right, height)
            if abs(left - right) > 1:
                self.isHeightBalanced = False
            return max(left, right)

        dfs(root)
        return self.isHeightBalanced

File: leetcode/leetcode_110.js
const isBalanced = (root) => {
  let isHeightBalanced = true;

  const dfs = (node, height = 0) => {
    if (!node) {
      return height;
    }
    height++;
    let [left, right] = [dfs(node.left, height), dfs(node.right, height)];
    if (Math.abs(left - right) > 1) {
      isHeightBalanced = false;
    }
    return Math.max(left, right);
  };
  dfs(root);
  return isHeightBalanced;
};
